Reworking logging in mreg.

terjekv opened the pull request:
This change migrates the entire logging backend to structlog (https://www.structlog.org/en/stable/) and implements an internal framework to offer the following funtionality:

  - Logs are context aware, and context is tracked via a request_id.
  - Each new request creates a new request_id.
  - Associated log entries are given the same request_id.
  - This request_id is also returned to the client via the header X-Request-ID.
  - Collation-IDs may be passed via X-Collation-ID and are both logged and returned to the client as such.
  - Console output is colorized via Rich
  - If the environment variable CI isn't set, the console gets raw JSON like the file.
  - The environment variable MREG_LOG_LEVEL controls the logging output.
  - To see an example in action, try `CI=True MREG_LOG_LEVEL=debug pytest mreg/api/v1/tests/test_hostgroups.py -vv -s`, or see https://gist.github.com/terjekv/8e97915c8b66c30241ff421a087e510d

Logging to file is also supported:

  - The log file gets raw json, no colours or sorting, but they inherit request_id and collation_id as expected.
  - The log file is rotated based on size (MREG_LOG_FILE_SIZE, defaults to 10MB) and a number of version are kept (MREG_LOG_FILE_COUNT, defaults to 5).
  - File name can be set by MREG_LOG_FILE_NAME (may contain slashes)
  - Log directory is extrated from the filename, and created if needed.

An example of the output can be seen here:
https://github.com/unioslo/mreg/pull/515#issuecomment-1647920124

terjekv commented:
<img width="1775" alt="Mreg logs 2023-07-24 at 15 31 31" src="https://github.com/unioslo/mreg/assets/656398/55b9abd4-1b5c-4bf2-9289-73fa9cd17146">

terjekv commented:
Note that there is no review or filtering of the content returned from http requests, nor on the contents offered in POST data.

terjekv commented on mregsite/settings.py:
```diff
@@ -197,13 +197,74 @@
+    cache_logger_on_first_use=True,
+)
+
+LOGGING = {
+    "version": 1,
+    "disable_existing_loggers": True,
+    "formatters": {
+        "console": {
+            "()": structlog.stdlib.ProcessorFormatter,
+            "processor": structlog.dev.ConsoleRenderer(),
+        },
+    },
+    "handlers": {
+        "console": {
+            "class": "logging.StreamHandler",
+            "formatter": "console",
+            "level": "DEBUG",
+        },
+    },
+    "loggers": {
+        "mreg": {
+            "handlers": ["console"],
+            "level": MREG_LOG_LEVEL,
+            "propagate": False,
+        },
+    },
+}
```
That makes sense. Although, with mreg running in containers I assume we'll mount a logs directory somewhere? For rotation, we can use RotatingFileHandler, akin to this (some tuning required):
````python
    'handlers': {
        'debugRotatingFileHandler': {
            'level':'DEBUG',
            'class':'logging.handlers.RotatingFileHandler',
            'formatter': 'verbose',
            'maxBytes': 1024*1024*1000, # 1000 MB
            'backupCount': 5, 
            'filename': os.path.join(BASE_DIR, 'logs/mreg.log'),
        },
    },

````

oyvindhagberg replied:
Yes, in the current deployments we mount a log directory under /app/logs in the container.
The name `logs` [comes from django-logging](https://github.com/cipriantarta/django-logging/blob/033db3f8c117788c97d172d6802e273f33de8dde/django_logging/settings.py#L35):
```python
self.__settings['LOG_PATH'] = os.path.join(str(django_settings.BASE_DIR), 'logs')
```
How about 100MB as a limit. That compresses to about 10MB. Easy to handle.

terjekv replied:
Added in https://github.com/unioslo/mreg/pull/515/commits/82c4dc32ed1b6588c2d4b5b45965594ec70de185.

oyvindhagberg commented:
I tested the container image from the artifact in https://github.com/unioslo/mreg/actions/runs/5961486496. I got it to log something to the console after setting `MREG_LOG_LEVEL=INFO`. It also created `logs/app.log`, but the log file remained empty... Any ideas?

terjekv commented:
I'm willing to bet that the path to the file is broken. Fixed a bug and added some hard validation of the name: https://github.com/unioslo/mreg/compare/928305a..61ea33a

oyvindhagberg commented:
When testing the latest changes locally, I get no log file at all. Tried both with and without setting `CI`.
I'm running `docker compose up` with the following changes to `docker-compose.yml`:
```yaml
+      - MREG_LOG_LEVEL=INFO
+      #- CI=yes
+    volumes:
+      - /mnt/oyvind/mreg/mregsite:/app/mreg/mregsite
+      - /mnt/oyvind/mreg/logs:/app/mreg/logs
```

terjekv commented:
Weird. Does it work if you run it with pytest or similar directly, ie, outside of a container?

terjekv commented:
Huh. It creates the logs folder, but not the file? That's really odd.

terjekv commented:
Right, this seems to be a bind issue. If I do `docker exec -it /bin/sh mreg-mreg` and `cat /app/logs/app.log` I see data in the log file.

oyvindhagberg commented:
Good news, it turns out I had used the wrong path for the volume. It worked after I corrected it.

terjekv commented:
At last it ensured I added some more checking and validation. :)
